pkg/s3repository/repository_protected_entity.go

```diff
@@ -141,37 +197,304 @@ func (this ProtectedEntity) GetID() astrolabe.ProtectedEntityID {
+			curBuf := discardBuf
+			if bytesToSkip < int64(len(discardBuf)) {
+				curBuf = discardBuf[0:bytesToSkip]
+			}
+			bytesRead, err := reader.Read(curBuf)
+			if err != nil && err != io.EOF {
+				return bytesSkipped, errors.Wrap(err, "Read failed while skipping")
+			}
+			bytesSkipped += int64(bytesRead)
+		}
+	}
+	return bytesSkipped, nil
+}
+func segmentName(baseName string, partNumber int, startOffset int64) string {
+	return baseName + "/" + fmt.Sprintf("%06d-%016d", partNumber, startOffset)
+}
+
+func parseSegmentName(key string) (baseName string, partNumber int, startOffset int64) {
+	lastSep := strings.LastIndexByte(key, '/')
+	if lastSep < 0 {
+		return "", -1, -1
+	}
+	baseName = key[0:lastSep]
+	segmentName := key[lastSep+1 : len(key)]
+	fmt.Sscanf(segmentName, "%d-%d", &partNumber, &startOffset)
+	return
+}
+
+const kMinMultiPartSize = 5 * 1024 * 1024
+const kMaxParts = 10000                 // Maximum number of parts we will use
+const kMaxBufferSize = 10 * 1024 * 1024 // Let's not get too crazy with buffers
+const kSegmentSizeLimit = kMaxParts * kMaxBufferSize
+
+func (this *ProtectedEntity) uploadSegment(ctx context.Context, baseName string, part int, startOffset int64, maxSegmentSize int64, reader io.Reader) (int64, error) {
+	if maxSegmentSize > kSegmentSizeLimit {
+		return 0, errors.Errorf("maxSegmentSize %d too large (%d is kSegmentSizeLimit)", maxSegmentSize, kSegmentSizeLimit)
+	}
+	name := segmentName(baseName, part, startOffset)
+	awsBucketName := aws.String(this.rpetm.bucket)
+	awsKey := aws.String(name)
+	completedParts := make([]*s3.Part, kMaxParts)
+
+	// First, check to see if there's an on-going multipart upload.  We assume that there will not be multiple
+	// repository managers trying to upload simultaneously, so if we find a multipart upload for this key, it needs
+	// to be restarted
+	listUploadsInput := &s3.ListMultipartUploadsInput{
+		Bucket: awsBucketName,
+		Prefix: awsKey,
+	}
+
+	multiPartUploadsOutput, err := this.rpetm.s3.ListMultipartUploadsWithContext(ctx, listUploadsInput)
+
+	var uploadID string
+	if len(multiPartUploadsOutput.Uploads) > 0 {
+		if len(multiPartUploadsOutput.Uploads) == 1 {
+			listPartsInput := &s3.ListPartsInput{
+				Bucket:           awsBucketName,
+				Key:              awsKey,
+				MaxParts:         nil,
+				PartNumberMarker: nil,
+				RequestPayer:     nil,
+				UploadId:         multiPartUploadsOutput.Uploads[0].UploadId,
+			}
+			err = this.rpetm.s3.ListPartsPagesWithContext(ctx, listPartsInput, func(listPartsOutput *s3.ListPartsOutput, b bool) bool {
+				for _, curPart := range listPartsOutput.Parts {
+					completedParts[int(*curPart.PartNumber)-1] = curPart
+				}
+
+				return true
+			})
+			if err != nil {
+				return 0, errors.Wrap(err, "ListPartsPagesWithContext failed")
+			}
+			uploadID = *multiPartUploadsOutput.Uploads[0].UploadId
+		} else {
+			// TODO - remove all the multipart uploads for this key something is wonky, should only have one in-flight
```

Comment:
is it better to log some message or return with error, even though it should be very unlikely to happen?